src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -258,7 +255,16 @@ public static DocumentAttributes ReadFrom(ObjectReader reader)
             }
 
             public Checksum Checksum
-                => _lazyChecksum ??= Checksum.Create(this, static (@this, writer) => @this.WriteTo(writer));
+            {
+                get
```

Comment:
📝 I'm working on a generic lock-free approach to present here

Reply:
Usage example:

```cs
return InterlockedOperations.Initialize(
    ref _lazyChecksum,
    ref _lazyChecksumInitialized,
    () => Checksum.Create(this, static (@this, writer) => @this.WriteTo(writer));
```

Reply:
It would be trivial to add a second overload that allows passing `this` as an argument to avoid the capture.

Reply:
i'm intersted in this.  but i'd prefer something *simple* for 17.9.  let's go with the simple lock (or a Lazy<>) for 17.9, and explore this lighter-weight interlocked-value-system for 17.10.   Deal? :)

Reply:
@CyrusNajmabadi Yes, I can wrap it, which also makes it the same size as `Nullable<T>`.

Reply:
i love it.  let's go with that.  and we can do that in 17.10.  thanks.

Reply:
It'd be interesting if the runtime could add such a helper directly, such that it was able to use the existing "is null" field in Nullable<T>.